File: src/precalculus_polar_coordinates_easy.py
"""
precalculus - polar_coordinates (easy)
"""

from problem_utils import *

# Full set of standard angles for varied theta values
THETA_OPTS = [
    0, pi/6, pi/4, pi/3, pi/2,
    2*pi/3, 3*pi/4, 5*pi/6, pi,
    7*pi/6, 5*pi/4, 4*pi/3, 3*pi/2,
    5*pi/3, 7*pi/4, 11*pi/6,
]

# Rectanglar points with clean polar forms (x, y, r, theta)
RECT_POLAR = [
    (1,  0,  1, 0),
    (0,  1,  1, pi/2),
    (-1, 0,  1, pi),
    (0, -1,  1, 3*pi/2),
    (1,  1,  sqrt(2), pi/4),
    (-1, 1,  sqrt(2), 3*pi/4),
    (-1,-1,  sqrt(2), 5*pi/4),
    (1, -1,  sqrt(2), 7*pi/4),
    (sqrt(3), 1, 2, pi/6),
    (1, sqrt(3), 2, pi/3),
    (-1, sqrt(3), 2, 2*pi/3),
    (-sqrt(3), 1, 2, 5*pi/6),
    (-sqrt(3),-1, 2, 7*pi/6),
    (-1,-sqrt(3), 2, 4*pi/3),
    (1, -sqrt(3), 2, 5*pi/3),
    (sqrt(3),-1, 2, 11*pi/6),
]


def generate():
    problem_type = randint(1, 6)

    if problem_type == 1:
        # Convert polar -> rectangular: (r, theta) -> (x, y)
        r_val = randint(1, 10)
        theta_val = choice(THETA_OPTS)
        x_val = simplify(r_val * cos(theta_val))
        y_val = simplify(r_val * sin(theta_val))
        ans = fmt_tuple([x_val, y_val])
        return problem(
            question=f"Convert the polar coordinates $\\left({r_val},\\,{latex(theta_val)}\\right)$ to rectangular $(x,y)$.",
            answer=ans,
            difficulty=(1000, 1150),
            topic="precalculus/polar_coordinates",
            solution=steps(
                "$x = r\\cos\\theta$, $y = r\\sin\\theta$",
                f"$x = {r_val}\\cos\\!\\left({latex(theta_val)}\\right) = {latex(x_val)}$",
                f"$y = {r_val}\\sin\\!\\left({latex(theta_val)}\\right) = {latex(y_val)}$",
                f"$(x,y) = ({latex(x_val)},\\,{latex(y_val)})$"
            ),
            answer_type="tuple"
        )

    elif problem_type == 2:
        # Convert rectangular -> polar: find r and theta
        rec = choice(RECT_POLAR)
        x_v, y_v, r_v, theta_v = rec
        # Scale r by random integer 1-6
        scale = randint(1, 6)
        x_val = simplify(scale * x_v)
        y_val = simplify(scale * y_v)
        r_val = simplify(scale * r_v)

        ask = choice(['r', 'theta', 'both'])
        if ask == 'r':
            ans = r_val
            return problem(
                question=f"Find $r$ for the rectangular point $({latex(x_val)},\\,{latex(y_val)})$.",
                answer=ans,
                difficulty=(1050, 1200),
                topic="precalculus/polar_coordinates",
                solution=steps(
                    "$r = \\sqrt{x^2 + y^2}$",
                    f"$r = \\sqrt{{({latex(x_val)})^2 + ({latex(y_val)})^2}} = {latex(r_val)}$"
                )
            )
        elif ask == 'theta':
            ans = theta_v
            return problem(
                question=f"Find $\\theta \\in [0, 2\\pi)$ for the rectangular point $({latex(x_val)},\\,{latex(y_val)})$.",
                answer=ans,
                difficulty=(1100, 1250),
                topic="precalculus/polar_coordinates",
                solution=steps(
                    "$\\theta = \\operatorname{atan2}(y, x)$, adjusting for quadrant",
                    f"$\\theta = {latex(theta_v)}$"
                )
            )
        else:
            ans = fmt_tuple([r_val, theta_v])
            return problem(
                question=f"Convert $({latex(x_val)},\\,{latex(y_val)})$ to polar form $(r,\\theta)$ with $r>0$, $0\\le\\theta<2\\pi$.",
                answer=ans,
                difficulty=(1100, 1250),
                topic="precalculus/polar_coordinates",
                solution=steps(
                    f"$r = {latex(r_val)}$",
                    f"$\\theta = {latex(theta_v)}$",
                    f"$(r,\\theta)=({latex(r_val)},\\,{latex(theta_v)})$"
                ),
                answer_type="tuple"
            )

    elif problem_type == 3:
        # Identify shape from simple polar equation
        eq_type = choice(['circle_r', 'ray', 'circle_cos', 'circle_sin', 'line_x', 'line_y'])

        if eq_type == 'circle_r':
            r_val = randint(1, 8)
            return problem(
                question=f"What shape does the polar equation $r = {r_val}$ describe?",
                answer="circle",
                difficulty=(1000, 1100),
                topic="precalculus/polar_coordinates",
                solution=steps(
                    f"$r = {r_val}$ means all points are at distance ${r_val}$ from the origin",
                    f"This is a circle of radius ${r_val}$ centered at the origin"
                ),
                answer_type="expression"
            )
        elif eq_type == 'ray':
            theta_val = choice([pi/6, pi/4, pi/3, pi/2, 2*pi/3, 3*pi/4, 5*pi/6])
            return problem(
                question=f"What shape does $\\theta = {latex(theta_val)}$ describe?",
                answer="line",
                difficulty=(1050, 1150),
                topic="precalculus/polar_coordinates",
                solution=steps(
                    "A constant angle $\\theta = c$ means every point lies on a ray through the origin",
                    "This describes a line (ray) through the origin"
                ),
                answer_type="expression"
            )
        elif eq_type == 'circle_cos':
            a_val = choice([2, 4, 6, 8])
            return problem(
                question=f"What shape does $r = {a_val}\\cos\\theta$ describe?",
                answer="circle",
                difficulty=(1100, 1250),
                topic="precalculus/polar_coordinates",
                solution=steps(
                    f"Multiply by $r$: $r^2 = {a_val}r\\cos\\theta$",
                    f"Substitute $r^2=x^2+y^2$, $r\\cos\\theta=x$: $x^2+y^2={a_val}x$",
                    f"Complete the square: $(x-{a_val//2})^2+y^2={a_val//2}^2$",
                    "This is a circle"
                ),
                answer_type="expression"
            )
        elif eq_type == 'circle_sin':
            a_val = choice([2, 4, 6, 8])
            return problem(
                question=f"What shape does $r = {a_val}\\sin\\theta$ describe?",
                answer="circle",
                difficulty=(1100, 1250),
                topic="precalculus/polar_coordinates",
                solution=steps(
                    f"Multiply by $r$: $r^2 = {a_val}r\\sin\\theta$",
                    f"$x^2+y^2={a_val}y$",
                    f"$(x)^2+(y-{a_val//2})^2={a_val//2}^2$: circle"
                ),
                answer_type="expression"
            )
        elif eq_type == 'line_x':
            a_val = randint(1, 6)
            return problem(
                question=f"Convert $r\\cos\\theta = {a_val}$ to rectangular form.",
                answer=Eq(x, a_val),
                difficulty=(1050, 1200),
                topic="precalculus/polar_coordinates",
                solution=steps(
                    "$r\\cos\\theta = x$",
                    f"$x = {a_val}$"
                ),
                answer_type="equation"
            )
        else:
            a_val = randint(1, 6)
            return problem(
                question=f"Convert $r\\sin\\theta = {a_val}$ to rectangular form.",
                answer=Eq(y, a_val),
                difficulty=(1050, 1200),
                topic="precalculus/polar_coordinates",
                solution=steps(
                    "$r\\sin\\theta = y$",
                    f"$y = {a_val}$"
                ),
                answer_type="equation"
            )

    elif problem_type == 4:
        # Evaluate r at given theta for a simple polar curve
        a_val = randint(1, 8)
        theta_val = choice([0, pi/6, pi/4, pi/3, pi/2, 2*pi/3, 3*pi/4, pi])
        curve = choice(['cos', 'sin', 'const_plus_cos'])

        if curve == 'cos':
            r_result = simplify(a_val * cos(theta_val))
            return problem(
                question=f"For $r = {a_val}\\cos\\theta$, find $r$ when $\\theta = {latex(theta_val)}$.",
                answer=r_result,
                difficulty=(1000, 1100),
                topic="precalculus/polar_coordinates",
                solution=steps(
                    f"$r = {a_val}\\cos\\!\\left({latex(theta_val)}\\right) = {a_val} \\cdot {latex(simplify(cos(theta_val)))} = {latex(r_result)}$"
                )
            )
        elif curve == 'sin':
            r_result = simplify(a_val * sin(theta_val))
            return problem(
                question=f"For $r = {a_val}\\sin\\theta$, find $r$ when $\\theta = {latex(theta_val)}$.",
                answer=r_result,
                difficulty=(1000, 1100),
                topic="precalculus/polar_coordinates",
                solution=steps(
                    f"$r = {a_val}\\sin\\!\\left({latex(theta_val)}\\right) = {a_val} \\cdot {latex(simplify(sin(theta_val)))} = {latex(r_result)}$"
                )
            )
        else:
            b_val = randint(1, 5)
            r_result = simplify(a_val + b_val * cos(theta_val))
            return problem(
                question=f"For $r = {a_val} + {b_val}\\cos\\theta$, find $r$ when $\\theta = {latex(theta_val)}$.",
                answer=r_result,
                difficulty=(1050, 1200),
                topic="precalculus/polar_coordinates",
                solution=steps(
                    f"$r = {a_val} + {b_val}\\cos\\!\\left({latex(theta_val)}\\right) = {a_val} + {b_val}\\cdot{latex(simplify(cos(theta_val)))} = {latex(r_result)}$"
                )
            )

    elif problem_type == 5:
        # Negative r: plot the equivalent positive-r representation
        r_val = randint(2, 8)
        theta_val = choice(THETA_OPTS)
        neg_theta = simplify(theta_val + pi)
        if neg_theta >= 2 * pi:
            neg_theta = simplify(neg_theta - 2 * pi)

        return problem(
            question=(f"The polar point $(-{r_val},\\,{latex(theta_val)})$ is the same as "
                      f"$({r_val},\\,\\theta)$ for what $\\theta \\in [0, 2\\pi)$?"),
            answer=neg_theta,
            difficulty=(1100, 1250),
            topic="precalculus/polar_coordinates",
            solution=steps(
                "A negative $r$ flips the point through the origin",
                f"$(-{r_val}, {latex(theta_val)}) = ({r_val}, {latex(theta_val)} + \\pi)$",
                f"$= ({r_val}, {latex(neg_theta)})$"
            )
        )

    else:
        # Compute r^2 and theta (easy angle) for a given rectangular point
        x_val = randint(1, 8)
        y_val = randint(1, 8)
        r_sq = x_val**2 + y_val**2
        return problem(
            question=f"For the rectangular point $({x_val},\\,{y_val})$, find $r^2$.",
            answer=r_sq,
            difficulty=(1000, 1100),
            topic="precalculus/polar_coordinates",
            solution=steps(
                "$r^2 = x^2 + y^2$",
                f"$r^2 = {x_val}^2 + {y_val}^2 = {x_val**2} + {y_val**2} = {r_sq}$"
            )
        )


emit(generate())

File: src/precalculus_polar_coordinates_easy.jl
# precalculus - polar_coordinates (easy)
# Generated: 2026-03-08T20:33:01.931066

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script r θ begin
    set_topic!("precalculus/polar_coordinates")
    
    problem_type = choice(1:5)
    
    if problem_type == 1
        # Convert polar to rectangular
        r_val = randint(2, 12)
        θ_vals = [0, π/6, π/4, π/3, π/2, 2π/3, 3π/4, 5π/6, π, 7π/6, 5π/4, 4π/3, 3π/2, 5π/3, 7π/4, 11π/6]
        θ_val = choice(θ_vals)
        
        x_ans = r_val * cos(θ_val)
        y_ans = r_val * sin(θ_val)
        
        # Simplify to exact values
        x_rounded = round(x_ans, digits=10)
        y_rounded = round(y_ans, digits=10)
        
        θ_str = if θ_val == 0
            "0"
        elseif θ_val == π
            "\\pi"
        elseif θ_val == π/2
            "\\frac{\\pi}{2}"
        elseif θ_val == 3π/2
            "\\frac{3\\pi}{2}"
        elseif θ_val == π/3
            "\\frac{\\pi}{3}"
        elseif θ_val == π/4
            "\\frac{\\pi}{4}"
        elseif θ_val == π/6
            "\\frac{\\pi}{6}"
        elseif θ_val == 2π/3
            "\\frac{2\\pi}{3}"
        elseif θ_val == 3π/4
            "\\frac{3\\pi}{4}"
        elseif θ_val == 5π/6
            "\\frac{5\\pi}{6}"
        elseif θ_val == 4π/3
            "\\frac{4\\pi}{3}"
        elseif θ_val == 5π/4
            "\\frac{5\\pi}{4}"
        elseif θ_val == 7π/6
            "\\frac{7\\pi}{6}"
        elseif θ_val == 5π/3
            "\\frac{5\\pi}{3}"
        elseif θ_val == 7π/4
            "\\frac{7\\pi}{4}"
        else
            "\\frac{11\\pi}{6}"
        end
        
        problem(
            question="Convert the polar coordinates \\(($(r_val), $(θ_str))\\) to rectangular coordinates \\((x, y)\\).",
            answer=fmt_tuple([x_rounded, y_rounded]),
            difficulty=(700, 900),
            answer_type="tuple",
            solution=steps(
                "Given: \\(r = $(r_val), \\theta = $(θ_str)\\)",
                "Use \\(x = r\\cos(\\theta)\\) and \\(y = r\\sin(\\theta)\\)",
                "\\(x = $(r_val)\\cos($(θ_str)) = $(x_rounded)\\)",
                "\\(y = $(r_val)\\sin($(θ_str)) = $(y_rounded)\\)",
                sol("Answer", fmt_tuple([x_rounded, y_rounded]))
            ),
            time=90,
            calculator="scientific"
        )
        
    elseif problem_type == 2
        # Convert rectangular to polar (r > 0)
        x_val = nonzero(-10, 10)
        y_val = nonzero(-10, 10)
        
        r_ans = sqrt(x_val^2 + y_val^2)
        θ_ans = atan(y_val, x_val)
        if θ_ans < 0
            θ_ans += 2π
        end
        
        r_rounded = round(r_ans, digits=4)
        θ_rounded = round(θ_ans, digits=4)
        
        problem(
            question="Convert the rectangular coordinates \\(($(x_val), $(y_val))\\) to polar coordinates \\((r, \\theta)\\) where \\(r > 0\\) and \\(0 \\leq \\theta < 2\\pi\\).",
            answer=fmt_tuple([r_rounded, θ_rounded]),
            difficulty=(800, 1000),
            answer_type="tuple",
            solution=steps(
                "Given: \\(x = $(x_val), y = $(y_val)\\)",
                "Calculate \\(r = \\sqrt{x^2 + y^2} = \\sqrt{$(x_val^2) + $(y_val^2)} = $(r_rounded)\\)",
                "Calculate \\(\\theta = \\arctan\\left(\\frac{y}{x}\\right)\\) with quadrant adjustment",
                "\\(\\theta \\approx $(θ_rounded)\\) radians",
                sol("Answer", fmt_tuple([r_rounded, θ_rounded]))
            ),
            time=120,
            calculator="scientific"
        )
        
    elseif problem_type == 3
        # Distance between two polar points
        r1 = randint(2, 10)
        r2 = randint(2, 10)
        θ_angles = [0, π/6, π/4, π/3, π/2, 2π/3, 3π/4, π]
        θ1 = choice(θ_angles)
        θ2 = choice(θ_angles)
        while θ1 == θ2
            θ2 = choice(θ_angles)
        end
        
        dist = sqrt(r1^2 + r2^2 - 2*r1*r2*cos(θ2 - θ1))
        dist_rounded = round(dist, digits=4)
        
        θ1_str = θ1 == 0 ? "0" : θ1 == π ? "\\pi" : θ1 == π/2 ? "\\frac{\\pi}{2}" : θ1 == π/3 ? "\\frac{\\pi}{3}" : θ1 == π/4 ? "\\frac{\\pi}{4}" : θ1 == π/6 ? "\\frac{\\pi}{6}" : θ1 == 2π/3 ? "\\frac{2\\pi}{3}" : "\\frac{3\\pi}{4}"
        θ2_str = θ2 == 0 ? "0" : θ2 == π ? "\\pi" : θ2 == π/2 ? "\\frac{\\pi}{2}" : θ2 == π/3 ? "\\frac{\\pi}{3}" : θ2 == π/4 ? "\\frac{\\pi}{4}" : θ2 == π/6 ? "\\frac{\\pi}{6}" : θ2 == 2π/3 ? "\\frac{2\\pi}{3}" : "\\frac{3\\pi}{4}"
        
        problem(
            question="Find the distance between the two points with polar coordinates \\(($(r1), $(θ1_str))\\) and \\(($(r2), $(θ2_str))\\).",
            answer=dist_rounded,
            difficulty=(900, 1100),
            solution=steps(
                "Use the polar distance formula: \\(d = \\sqrt{r_1^2 + r_2^2 - 2r_1r_2\\cos(\\theta_2 - \\theta_1)}\\)",
                "Substitute: \\(d = \\sqrt{$(r1)^2 + $(r2)^2 - 2($(r1))($(r2))\\cos($(θ2_str) - $(θ1_str))}\\)",
                "Calculate: \\(\\cos($(θ2_str) - $(θ1_str)) = $(round(cos(θ2 - θ1), digits=4))\\)",
                "\\(d = \\sqrt{$(r1^2) + $(r2^2) - $(round(2*r1*r2*cos(θ2 - θ1), digits=4))}\\)",
                sol("Answer", dist_rounded)
            ),
            time=150,
            calculator="scientific"
        )
        
    elseif problem_type == 4
        # Convert polar equation to rectangular
        eq_type = choice(1:3)
        
        if eq_type == 1
            # r = k (circle centered at origin)
            k = randint(2, 15)
            problem(
                question="Convert the polar equation \\(r = $(k)\\) to rectangular form.",
                answer="x^2+y^2=$(k^2)",
                difficulty=(700, 900),
                solution=steps(
                    "Given: \\(r = $(k)\\)",
                    "Use \\(r^2 = x^2 + y^2\\)",
                    "Square both sides: \\(r^2 = $(k^2)\\)",
                    sol("Answer", "x^2 + y^2 = $(k^2)")
                ),
                time=60
            )
        elseif eq_type == 2
            # θ = k (line through origin)
            θ_vals_simple = [0, π/6, π/4, π/3, π/2, 2π/3, 3π/4, 5π/6, π]
            θ_val = choice(θ_vals_simple)
            
            θ_str = θ_val == 0 ? "0" : θ_val == π ? "\\pi" : θ_val == π/2 ? "\\frac{\\pi}{2}" : θ_val == π/3 ? "\\frac{\\pi}{3}" : θ_val == π/4 ? "\\frac{\\pi}{4}" : θ_val == π/6 ? "\\frac{\\pi}{6}" : θ_val == 2π/3 ? "\\frac{2\\pi}{3}" : θ_val == 3π/4 ? "\\frac{3\\pi}{4}" : "\\frac{5\\pi}{6}"
            
            if θ_val == 0
                ans = "y=0"
            elseif θ_val == π
                ans = "y=0"
            elseif θ_val == π/2
                ans = "x=0"
            else
                m = tan(θ_val)
                m_rounded = round(m, digits=4)
                ans = "y=$(m_rounded)x"
            end
            
            problem(
                question="Convert the polar equation \\(\\theta = $(θ_str)\\) to rectangular form.",
                answer=ans,
                difficulty=(800, 1000),
                solution=steps(
                    "Given: \\(\\theta = $(θ_str)\\)",
                    "Use \\(\\tan(\\theta) = \\frac{y}{x}\\)",
                    "\\(\\tan($(θ_str)) = \\frac{y}{x}\\)",
                    sol("Answer", ans)
                ),
                time=90,
                calculator="scientific"
            )
        else
            # r = a cos(θ) or r = a sin(θ)
            a = nonzero(-8, 8)
            is_cos = choice([true, false])
            
            if is_cos
                # r = a cos(θ) => x^2 + y^2 = ax => (x-a/2)^2 + y^2 = (a/2)^2
                problem(
                    question="Convert the polar equation \\(r = $(a)\\cos(\\theta)\\) to rectangular form.",
                    answer="x^2+y^2=$(a)x",
                    difficulty=(900, 1100),
                    solution=steps(
                        "Given: \\(r = $(a)\\cos(\\theta)\\)",
                        "Multiply both sides by \\(r\\): \\(r^2 = $(a)r\\cos(\\theta)\\)",
                        "Use \\(r^2 = x^2 + y^2\\) and \\(x = r\\cos(\\theta)\\)",
                        sol("Answer", "x^2 + y^2 = $(a)x")
                    ),
                    time=90
                )
            else
                # r = a sin(θ) => x^2 + y^2 = ay
                problem(
                    question="Convert the polar equation \\(r = $(a)\\sin(\\theta)\\) to rectangular form.",
                    answer="x^2+y^2=$(a)y",
                    difficulty=(900, 1100),
                    solution=steps(
                        "Given: \\(r = $(a)\\sin(\\theta)\\)",
                        "Multiply both sides by \\(r\\): \\(r^2 = $(a)r\\sin(\\theta)\\)",
                        "Use \\(r^2 = x^2 + y^2\\) and \\(y = r\\sin(\\theta)\\)",
                        sol("Answer", "x^2 + y^2 = $(a)y")
                    ),
                    time=90
                )
            end
        end
        
    else
        # Plot a point in polar coordinates (conceptual)
        r_val = randint(1, 8)
        θ_vals = [π/6, π/4, π/3, π/2, 2π/3, 3π/4, 5π/6, π, 7π/6, 5π/4, 4π/3, 3π/2, 5π/3, 7π/4, 11π/6]
        θ_val = choice(θ_vals)
        
        x_coord = r_val * cos(θ_val)
        y_coord = r_val * sin(θ_val)
        
        x_rounded = round(x_coord, digits=4)
        y_rounded = round(y_coord, digits=4)
        
        θ_str = θ_val == π/6 ? "\\frac{\\pi}{6}" : θ_val == π/4 ? "\\frac{\\pi}{4}" : θ_val == π/3 ? "\\frac{\\pi}{3}" : θ_val == π/2 ? "\\frac{\\pi}{2}" : θ_val == 2π/3 ? "\\frac{2\\pi}{3}" : θ_val == 3π/4 ? "\\frac{3\\pi}{4}" : θ_val == 5π/6 ? "\\frac{5\\pi}{6}" : θ_val == π ? "\\pi" : θ_val == 7π/6 ? "\\frac{7\\pi}{6}" : θ_val == 5π/4 ? "\\frac{5\\pi}{4}" : θ_val == 4π/3 ? "\\frac{4\\pi}{3}" : θ_val == 3π/2 ? "\\frac{3\\pi}{2}" : θ_val == 5π/3 ? "\\frac{5\\pi}{3}" : θ_val == 7π/4 ? "\\frac{7\\pi}{4}" : "\\frac{11\\pi}{6}"
        
        problem(
            question="What are the rectangular coordinates of the point with polar coordinates \\(($(r_val), $(θ_str))\\)?",
            answer=fmt_tuple([x_rounded, y_rounded]),
            difficulty=(700, 900),
            answer_type="tuple",
            solution=steps(
                "Use x = r*cos(theta) and y = r*sin(theta) to convert",
                sol("Answer", fmt_tuple([x_rounded, y_rounded]))
            ),
            time=90
        )
    end
end
